### PR TITLE
Allow private image pulls

### PR DESCRIPTION
**Pull request by frodopwns**
This PR allows you to pass docker hub credentials to the watchtower image as ENV vars and thus find images that are private.

### CODE REVIEW

**Comment by frodopwns** [comment]
Looks like some functions in the tests need to be updated to reflect newly required parameters.

eg. RemoveImage() now requires *AuthConfig or nil

edit: on second thought this may be testing with the wrong version of dockerclient

**Comment by frodopwns** [comment]
Also i guess the change to ubuntu is a drastic one.  I did this because the watchtower binary would not run in scratch or alpine....possibly due to dynamic linking?

**Comment by phumpal** [comment]
@frodopwns very cool. 

I've tested this and it seems to work well with Docker Hub but doesn't seem to support images pulled from a private (self-hosted) repo.

**Comment by frodopwns** [comment]
@phumpal yeah it wasn't readily apparent to me in the https://github.com/samalba/dockerclient how you would pull from another registry.

**Comment by bahlo** [comment]
:+1:

**Comment by nafg** [comment]
What's the status of this?

**Comment by bahlo** [comment]
@frodopwns: I checked and I think, you need to set the `DockerClient.URL` property to change the registry. I've not tested this, but it is used in all registry operations.

I'm not sure though ¯\_(ツ)_/¯

File: container/client.go [hunk 1]
```diff
@@ -111,7 +116,19 @@ func (client dockerClient) StartContainer(c Container) error {
...
-	newContainerID, err := client.api.CreateContainer(config, name)
+	var err error
+	var newContainerID string
+	if username != "" && password != "" && email != "" {
+		auth := dockerclient.AuthConfig{
+			Username: username,
+			Password: password,
+			Email:    email,
+		}
+		newContainerID, err = client.api.CreateContainer(config, name, &auth)
```
Comment (bahlo) [inline container/client.go:127]: Maybe set `dockerclient.URL` manually to `os.Getenv("REPO_REGISTRY")`, if not empty, or something?

**Comment by frodopwns** [comment]
Looking into using other than Docker Hub repos with credentials...

**Comment by chrisferry** [comment]
What's the status of this?  Any chance it will support ECR Auth?

**Comment by rossvz** [comment]
Is there a timeline on this being merged?

**Comment by curry684** [comment]
The switch to Ubuntu makes this pretty much worthless for any serious use given the amount of disk space it's killing now if you don't use Ubuntu:14.04 in other containers. The point of a watchtower should pretty much be sitting unobtrusively in a very small (5MB) corner of the system methinks.

**Comment by frodopwns** [comment]
@curry684 that part probably isn't necessary. I just didn't want to use that previous image in my testing for some reason i can't remember.  This is still just a go app so it should run anywhere go binaries do.

**Comment by curry684** [comment]
I agree, but it still got merged.

**Comment by frodopwns** [comment]
@curry684 https://github.com/CenturyLinkLabs/watchtower/pull/21

**Comment by fdattein** [comment]
@frodopwns funny enough, changes in #21 aren't in master, despite github saying it has been merged. Am I missing something?

**Pull request merged**